### ISSUE TITLE
MEMORY_LIMIT should be an excluded environment variable.

### ISSUE THREAD

**Issue by nebhale** [cloudfoundry/cloud_controller_ng#832]
Currently, a user can manually set `$MEMORY_LIMIT` using `cf set-env`.  This variable is supposed to reflect the value that container believes is correct and is depended on by the Java Buildpack to be accurate.  When a user sets the value manually, the JVM started by the Java Buildpack can be the incorrect size (too large being worse than too small) and could lead to the application being killed by the container.

The `$MEMORY_LIMIT` environment variable shouldn't be user settable to prevent this ambiguity.  This should match the behavior of `$VCAP_*` variables.

**Comment by zrob** [comment]
@nebhale while we don't throw an error when a user sets MEMORY_LIMIT, we do override anything a user sets with the platform value.  So no user should be able to get into the state you are describing.

**Comment by nebhale** [comment]
@zrob Experimentally I've determined that user environment variables *do override* the container truth.  Happy to demonstrate in the office if needed.

**Comment by zrob** [comment]
@nebhale can you tell me which cf version you're using? i'm unable to reproduce that behavior. or we can just sync up on monday

**Comment by nebhale** [comment]
`cf ssh` was how I verified it, but it seems like you've gotten confirmation.

**Comment by tusing** [comment]
Upon further investigation, we can summarize our understanding of this issue as the following:

* Pushing an app
    * Ignores the platform memory limit

        ```bash
        $ cf set-env <app> MEMORY_LIMIT 900G
        $ cf restage <app>
        $ curl <app>/env | grep MEMORY_LIMIT
        
        "MEMORY_LIMIT"=>"900G" # app memory limit exceeds platform memory limit
        ```

* Pushing a task
    * Correctly adheres to the platform memory limit

        ```bash
        $ cf run-task a 'env' -m 900G
        Creating task for app a in org o / space s as admin...
        memory_in_mb exceeds organization memory quota
        FAILED
        ```
    

    * But ignores the the app memory limit


        ```bash
        $ cf set-env <app> MEMORY_LIMIT 256M
        $ cf restage <app>
        $ curl <app>/env | grep MEMORY_LIMIT
        "MEMORY_LIMIT"=>"256M" # app memory limit 256M
        
        $ cf run-task <app> 'env' -m 1G
        $ cf logs --recent <app> | grep MEMORY_LIMIT
        [APP/TASK/e262f158/0] OUT MEMORY_LIMIT=1024m # we don't adhere to the app memory limit of 256M
        ```

Proceeding with regards to pushing an app, do we want to

* Ignore the user-set limit for apps entirely, or

* Remove the ability for the user to set an app memory limit in the first place, or

* Override the platform limit with the user limit for an app *iff* the user limit is less than the platform limit

With regards to pushing a task, we are assuming that the behavior with the task memory limit overriding the app memory limit is unintended - is this correct?

**Comment by nebhale** [comment]
My vote is for this to be just like `VCAP_*`; that it is illegal to set it at all.  That way the user never has to reason about whether their environment variables is being used and how changing the app memory limit might affect that.

**Comment by ljfranklin** [comment]
@zrob sounds like we've been able to reproduce this inconsistency between App and Task behavior w.r.t. setting memory limits. Is this something you'd like the engineering team to look into or are we marking this as open for contribution?

**Comment by Gerg** [comment]
For now, let's make apps behave the same as tasks (system variables override user-provided variables). We can re-visit restricted variable names separately.

**Comment by nebhale** [comment]
Seems like a reasonable compromise.  They key is that the system needs to be the authority on this, telling the user why their configuration is being overridden is icing on the cake.

**Comment by lisamcho** [comment]
So as I see it the acceptance criteria should be...
Given: platform default_app_memory of 256

* Pushing an app
    * Platform overrides the user-set MEMORY_LIMIT
        ```bash
        $ cf set-env dora MEMORY_LIMIT 100M
        $ cf restage dora
        $ curl <app>/env | grep MEMORY_LIMIT
        
        "MEMORY_LIMIT"=>"256"
        ```

* Pushing a task the user-set $MEMORY_LIMIT env var of the parent app should be ignored**
        ```bash
        $ cf run-task dora 'env' -m 50M
OUT MEMORY_LIMIT=50M
        ```

**Note there is a separate issue we've filed as to whether the task should respect the platform set memory of the parent app (https://www.pivotaltracker.com/story/show/152085405)

**Comment by elenasharma** [comment]
We addressed this in d5e4d00c2df3f371c9dbdaf044b393f1a1c07033.
In the event that the user specifies a `MEMORY_LIMIT`, the `MEMORY_LIMIT` env var will still come from the process.
